tensorflow_model_optimization/g3doc/guide/clustering/index.md

```diff
@@ -0,0 +1,99 @@
+
+## Overview
+
+Clustering, or weight sharing, reduces the number of unique weight values in a model, leading to benefits for deployment. It first groups the weights of each layer into *N* clusters, then shares the cluster's centroid value for all the weights belonging to the cluster.
+
+This technique brings improvements in terms of model compression. By reducing the number of unique weight values, weight clustering renders the weights suitable for compression via Huffman coding and similar techniques. Future framework support will, therefore, be able to reduce memory size and bandwidth requirements. This can be critical for deploying deep learning models on embedded systems with limited resources.
+
+We have seen up to 5x improvements in model compression with minimal loss of accuracy, as demonstrated by the [results](#results) presented below. The compression gains depend on the model and the accuracy targets in each specific use case. For example, for the MobileNetV2 image classification model, one can choose to reduce all non-depthwise convolutional layers to use just 32 unique weigth values and obtain a float32 tflite model that is approximately 4.8 times more compressible using ZIP Deflate algorithm than the original model. However, that will result in about 3% drop of the top-1 classification accuracy. On the other hand, the same model clustered less agressively, using 256 clusters for two internal layers and 32 clusters for the final convolutional layer, maintains virtually the same accuracy as the original model, yet still yields a respectable 1.8x improvement in compression ratio.
+
+Clustering works well with TFLiteConverter, providing an easy path to produce deployment-ready models that can be easily compressed using either an off-the-shelf compression algorithm, similar to the ZIP Deflate we use for demonstration in this document, or a custom method optimized for a special target hardware. When converting the clustered model with TFLiteConverter, the actual number of unique weight values per tensor may increase. This happens for the models with batch normalization layers that are folded into the preceding convolutional layers during the conversion, and also due to different scale factors in the per-channel weight quantization scheme. Both techniques may alter the same weight value differently, depending on the channel it appears in and the associated batch-normalization and quantization parameters. While this side effect may result in a slightly lower compression ratio, the overall benefits of using clustering and post-training conversion and quantization are still tangible, as demonstrated by the examples in this document.
```

Comment:
For a user making a decision, a lot of these details don't help with deciding whether to use weight clustering. 

"TFLiteConverter, providing an easy path to produce deployment-ready models that can be easily compressed using either an off-the-shelf compression algorithm, similar to the ZIP Deflate"

This goes into the "how to use", which users should just reference the tutorials for. 

Something to the point for this whole paragraph is:
"Clustering will provide useful but reduced benefits for convolution and dense layers that precede a batch normalization layer and in combination with per-axis post-training quantization."

They can otherwise see the consequences from the Mobilenet v1 and v2 numbers in the below, that face the batchnorm folding issue. If you can quantify the consequences with a range (Nx - Mx instead of the usual Px - Qx), that'd help also.

The details (e.g. a specific one with the math) may be interesting, but in a "Internal" page for people who want to understand how things work. We haven't created this page yet for other techniques, but will need to once we move the quantization scheme over to the TFMOT docs. 

@Ruomei